test/src/test/integration/git_protocol.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright © 2021 The Radicle Link Contributors
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::collections::BTreeSet;
+
+use futures::TryFutureExt as _;
+use radicle_link_git_protocol::{fetch, upload_pack};
+use tempfile::TempDir;
+
+fn upstream() -> TempDir {
+    let tmp = tempfile::tempdir().unwrap();
+
+    let repo = git2::Repository::init_bare(&tmp).unwrap();
+    let auth = git2::Signature::now("apollo", "apollo@creed.me").unwrap();
+
+    let tree = {
+        let empty = repo.treebuilder(None).unwrap();
+        let oid = empty.write().unwrap();
+        repo.find_tree(oid).unwrap()
+    };
+    let base = {
+        let oid = repo
+            .commit(
+                Some("refs/namespaces/foo/refs/heads/main"),
+                &auth,
+                &auth,
+                "initial",
+                &tree,
+                &[],
+            )
+            .unwrap();
+        repo.find_commit(oid).unwrap()
+    };
+    let next = repo
+        .commit(
+            Some("refs/namespaces/foo/refs/heads/next"),
+            &auth,
+            &auth,
+            "ng",
+            &tree,
+            &[&base],
+        )
+        .unwrap();
+    repo.reference(
+        "refs/namespaces/foo/refs/pulls/1/head",
+        next,
+        true,
+        "pee arrr",
+    )
+    .unwrap();
+
+    tmp
+}
+
+#[test]
+fn standard_clone() {
+    let upstream = upstream();
+
+    let (client_recv, server_send) = futures_ringbuf::Endpoint::pair(1024, 1024);
+    let (server_recv, client_send) = futures_ringbuf::Endpoint::pair(1024, 1024);
+
+    let client = async move {
+        let opt = fetch::Options {
+            repo: "foo".into(),
+            extra_params: vec![],
+            ref_prefixes: vec!["refs/heads/".into(), "refs/pulls/".into()],
+            want_refs: vec![],
+        };
+
+        fetch::fetch(
+            opt,
+            |r| {
+                Ok(Some(fetch::WantHave {
+                    want: *r.unpack().1,
+                    have: None,
+                }))
+            },
+            fetch::DiscardPack,
+            client_recv,
+            client_send,
+        )
+        .await
+    };
+    let server =
+        upload_pack::upload_pack(&upstream, server_recv, server_send).and_then(|(_hdr, run)| run);
+
+    let client_out = {
+        use futures::{
+            executor::block_on,
+            future::{select, Either},
+            pin_mut,
+        };
+
+        pin_mut!(client);
+        pin_mut!(server);
+        block_on(async move {
+            match select(client, server).await {
+                Either::Left((res, _)) => res,
```

Comment:
This is interesting: When waiting for the server and running the test it hangs even though the client is done. Here I would immediately assume that something is missing in the protocol, like a finishing empty request.

```suggestion
                Either::Left((res, server)) => {
                    server.await;
                    res
                }
```

Edit: _Thinking about it, it's a bit strange that the server still has a connection to listen on as `client_send` should already be dropped, stdin should also become unavailable for reading. Maybe related to async, or I am missing something_.

Having re-read the docs it became clear that [`gitoxide` is missing the empty request](https://github.com/git/git/blob/master/Documentation/technical/protocol-v2.txt#L136:L140) at the end of the operation. In V1 the session automatically ends after a packfile was received, but in V2 that's not the case anymore being a command-oriented protocol now. This is only an issue in persistent V2 connections as well like the ones used here.

I vaguely recall that the server shutting down isn't desired behaviour too and that sending a flush packet/empty request does indeed shut the server connection down. Could you elaborate on why that is, @kim ?

Reply:
I’m not sure actually if `futures_ringbuf` implements the dropping behaviour we’d expect here correctly. Perhaps a named pipe would be preferable for these kinds of scenarios, but the only one that turned up from a quick search was a semi-functional implementation in recent `tokio`.


To elaborate on the behaviour we want here: the stateful variants of the protocol are either `git-daemon` or SSH. Both imply some degree of trust between the parties. It is indeed the case that `link` could re-use the same connection (ie QUIC stream) to ask for a subsequent packfile (recall: we need to get at some signed data first, and then get the signed heads). However, we can not currently look into what’s going on from the server side, because we’re just shuffling bytes to `stdio` of `git-upload-pack`.

That creates a security concern: what if a client just does nothing, but keeps the connection (and thus, the upload-pack process) alive? It does not seem like the `—timeout` option has an effect in this case, but perhaps worth looking into why this is the case.

Since QUIC streams are (relatively) cheap, I would at this point prefer to run things in stateless mode, much like HTTP, but without the overhead of assembling HTTP headers. (_Aside: I have no idea how a multi-round negotiation could possibly work over HTTP_) Alternatively, we could wrap the receive stream and close it if either no bytes or too many bytes were received for some period of time, but that’s a lot more complicated. Ultimately, the “server” end will need to understand which commands are being sent in what order, and terminate the session if that doesn’t adhere to what a well-behaved `link` peer would do.

Reply:
Thank you, all noted. I take it that the `git-upload-pack` process should definitely shut down at the end of the interaction, but it not doing so may also be related to the ring-buffer not working as expected on drop. Maybe that calls for a test on process level so the connection is real, or… bring up a real tcp socket along with a real connection. Of course, one would still have to hope that `tokio` tcp connections act accordingly on drop.

I will also run a test or two to verify it's not `gitoxide` misbehaving, and I have a strong feeling it is _in some circumstances_ and this being one of them.

> That creates a security concern: what if a client just does nothing, but keeps the connection (and thus, the upload-pack process) alive?
> Ultimately, the “server” end will need to understand which commands are being sent in what order, and terminate the session if that doesn’t adhere to what a well-behaved link peer would do.

These look like excellent requirements for the server implementation :).

> (Aside: I have no idea how a multi-round negotiation could possibly work over HTTP)

This one _I think_ I know - the client sends all the wants and haves from the previous round. The `Arguments` type [is implementing this](https://github.com/Byron/gitoxide/blob/0d5b911ad5f47ab8f044d6bbe660a6d1dfeecb5f/git-protocol/src/fetch/arguments/async_io.rs#L35). For V2 it always does that as it's supposed to be stateless, in V1 it's optional depending on whether the transport says so (i.e. V1 over HTTP is stateless, hence needs to resend all previous haves and wants).

Reply:
I played around with this more and found:

* `--timeout` only [applies to v1](https://git.kernel.org/pub/scm/git/git.git/tree/builtin/upload-pack.c#n41)
* in v2 mode, `git-upload-pack` (without any involvement from `git-daemon`) will terminate the command loop if either [EOF](https://git.kernel.org/pub/scm/git/git.git/tree/serve.c#n217) or a [flush pkt](https://git.kernel.org/pub/scm/git/git.git/tree/serve.c#n236), and no command, is encountered
* in v2 mode, `--stateless-rpc` means that `git-upload-pack` will wait for exactly one command, process it, and then exit
* I still don't know why dropping one end of a `futures_ringbuf::Endpoint` pair doesn't lead to delivering EOF to `upload-pack`. If you drop `server` [here](https://github.com/najamelan/futures_ringbuf/blob/dev/examples/endpoint.rs#L25) and then try to read again, it works as expected.

So what I ended up with is to simply send a flush after we're done fetching (d4dd9d636cfc2416d25735196a369af6ded8d953). I conclude that this is the as-intended way to use v2 -- after all, we _do_ want to send "non-standard" command sequences eventually. Probably the flush should even be up to the caller.

I wonder where you see `gitoxide` misbehaving -- are you thinking of using a `Action` variant to decide whether to send a flush after a packfile was received?

Reply:
> I wonder where you see gitoxide misbehaving -- are you thinking of using a Action variant to decide whether to send a flush after a packfile was received?

I see what you mean. From my point of view the operation is done once `git_protocol::fetch(…)` is completed and seeing what happened here it would be the least surprising to send a flush packet automatically when it's finished. I would add a default-implemented method to the delegate to allow to override this if there are other plans, like fetching multiple different packs.